### ISSUE TITLE
Progress measures for grokking via mechanistic interpretability

### ISSUE THREAD

**Issue by Tonnonssi** [Tonnonssi/xRL#6]
## 🥅 목적

MI를 낋여보자~!!

## 🤩 공부할 내용
- 사랑이가 알려주겠지...... 뭔가 발견했는데! 엄청 중요하다! 개쩐다 ! 
- 

## 📄 참고 자료 
[Progress measures for grokking via mechanistic interpretability](https://arxiv.org/pdf/2301.05217)

**Comment by Tonnonssi** [comment]
## 초록 
emergent behavior, 창발적 행동 : 파라미터, 학습 데이터 셋, 학습 스텝 수를 늘렸을 때 발생하는 모델의 질적 향상
이런 창발적 행동을 이해하기 위해서는 이산적으로 보이는 질적 변화를 연속적으로 관측 및 해석하는 과정이 필요하다. 본 논문은 이걸 reverse-
engineering learned behaviors into their individual components로 접근한다.  
특별히 grokking(모델이 학습 데이터에 대해 과적합(Overfitting)되어 훈련 정확도는 완벽하지만 테스트 정확도는 바닥인 상태가 오랫동안 지속되다가, 어느 순간 갑자기 테스트 정확도가 급상승하며 일반화(Generalization)되는 현상: Power et al., 2022)을 분석했다.   
타겟 신경망 : modular addition을 수행하는 작은 트랜스포머 
사용한 방법론 : 이산 푸리에 변환과 trigonometric identities(삼각함수 항등식)으로 원 위의 회전으로 덧셈을 변환 (input : a,b > ft: (a+b) mod P = ? )
구체적 대상 : 활성화값과 가중치 -> 푸리에 공간에서 제거 실험 
학습 단계 : memorization, circuit formation, and cleanup 로 정의 
결론 : grokking은 갑자기 발생하는 현상이 아니라, 암기내용을 지우며 따라오는 가중치에 내재된 점진 확장이다.  

Phase Changes : 갑작스럽게 독자적으로 질적으로 높은 행동을 배우는 것 , 모델을 키우거나 더 오래 학습시켰을 때 발생 
## 서론 
emergent behavior은 자주 발생하는데, emergent reward hacking이라는 리스크를 가진다. 이거는 너무 **abrupt**(갑작스럽게) 발생한다는 문제점이 있다. 다른 논문들도 cross entropy loss를 이용해 이걸 해석하려는 모습을 보였지만, 근본적인 해석을 주지 못한다. 

### Reverse Engineering을 통해 찾아낸 것 

<img width="1388" height="566" alt="Image" src="https://github.com/user-attachments/assets/40dcfddc-9b3e-4b18-a07e-beee6faa00e2" />  
그림에 써 있는 단계 별로 모델이 수행하는 것을 확인했다.  
$w_k$(key frequencies) 는 모델이 입력값 a,b를 임베딩할 떄 사용하는 벡터로 sparse하다는 특정이 있다. 얘는 삼각함수 변환을 통해서 모듈러 덧셈 문제를 접근하고 해결한다! 
+ 일반 트랜스포머와 다른 이유 : 테스크가 간단하기 때문에 간단하게 만들었다. $h_0$은 어텐션이다. 

### 작동 근거 
1. 네트워크의 가중치와 활성화값이 일관적인 주기 구조를 가진다. 
2. $W_L$은 은 잘 근사된 $w_k$의 삼각함수의 합이다. 여기에 MLP로 활성화시키면 신경망이 지닌 삼각함수적 특징을 알 수 있다. 
3. 어텐션 헤드와 MLP가 2차 삼각함수라는 걸 확인 
4. $w_k$를 억제하면 성능이 나빠지는데, $w_k$를 제외한 나머지를 억제하면 성능이 오히려 좋아진다.  

### 테스크를 이해하기 위해 두 개의 Loss를 정의  
1. excluded loss : $w_k$를 제거하고 loss 계산 
2. restricted loss : $w_k$를 제외한 나머지를 제외하고 ㅣoss 계산  

두 지표모두 그로킹이 발생하기 앞서 나아짐  
지표 사용 방법 : 학습 단계를 이해하기 위해 사용함  

### 학습 단계 
1. 암기 : 학습 데이터를 외움
2. 회로 구축 : 신경망이 작동원리를 배우고, 일반화하는 과정 
3. 클린업 : 가중치 감소가 암기 내용을 지움 
일반화 이후 단계인 클린업 단계에서 그로킹이 발생  
> 그로킹은 갑작스러운 변화가 아니고, 암기한 거를 잊으며 발생하는 가중치에 내재되는 점진적 함축 과정에서 나온 것이다. 

## 학습 과정 
자세한 모델 스펙은 논문 참고하시길   
<img width="600" height="173" alt="Image" src="https://github.com/user-attachments/assets/20e8cee4-8715-445e-ba5f-a297be601d7f" />
전형적으로 그로킹 현상이 포착됨

**Comment by Tonnonssi** [comment]
## 정의한 푸리에 로직

<img width="678" height="92" alt="Image" src="https://github.com/user-attachments/assets/eacfdb68-3f72-41e2-b331-4acf4d8effb7" />

<img width="658" height="249" alt="Image" src="https://github.com/user-attachments/assets/35d79922-89d1-4946-b66c-b29bc70c61cb" />

## 🌟 4장 : REVERSE ENGINEERING
위에서 언급한 작동근거를 단계별로 제시한다.  

### 1. 주기성: 가중치와 활성화값에서 찾을 수 있다. 
> 가설 : 푸리에 변환과 L2 정규화를 취한 값과 신경망의 임베딩과 값이 동일할 것    

🔥 푸리에 변환의 핵심 : 원래 값으로 복원하는 것. 신호 분해     

$W_E$ : 신경망에서 임베딩된 행렬  
$W_L$ : 신경망의 로짓 맵
과정 : $W_E$에 푸리에 변환 후 L2 정규화 

<img width="665" height="304" alt="Image" src="https://github.com/user-attachments/assets/96aa30af-d1d1-4a45-864d-ee68c027ba18" />  

symmetric해서 P=113의 절반인 k=56만 이용. 
왼쪽 그래프는 $W_E$에서 푸리에 변환을, 오른쪽 그래프는 $W_L$ 에서 푸리에 변환을 적용한 뒤 norm 푸리에 구성요소다.  
두 그래프는 거의 유사한데, 이는 임베딩에도, 출력로짓층에도 동일한 $w_k$가 유지되고 있음을 의미한다. 또한 이건 엄청나게 sparse하다.  

<img width="621" height="190" alt="Image" src="https://github.com/user-attachments/assets/83566f88-483d-4916-a8dc-504848f3b569" />  

왼 : head 0의 ‘=’  ‘a’사이 어텐션 점수  
중간 : a,b가 들어왔을 때 MLP 활성화값  
-> 둘 다 주기성을 띔 (k=42가 기준)  
오 : 로짓의 푸리에 구성값의 norm : 5개의 핵심 주기에 상응하는 삼각함수 값만 활성화됨, 로짓은 주기성을 가짐    

### 2. 삼각함수의 공식을 배운다.  

<img width="656" height="129" alt="Image" src="https://github.com/user-attachments/assets/a9514f60-14d5-48d3-b364-be07533b80f2" />

<img width="673" height="207" alt="Image" src="https://github.com/user-attachments/assets/17070775-a844-492b-8047-bb8b1e63845d" />

<img width="647" height="69" alt="Image" src="https://github.com/user-attachments/assets/6ceda9b6-1ec4-490f-b1d2-3022c53581b7" />

이 과정을 통해 할 수 있는 것 :   
1. $W_L$이 $cos(w_k c)$와 $sin(w_k c)$으로 추론됨  
2. MLP가 cos(a+b)와 sin(a+b)를 잘 근사한다.  

### 3. MLP는 각 주기의 2차 함수를 배운다.  
2차 함수가 중요한 이유는 삼각함수의 합공식은 전부 cos과 sin 중 2개가 곱해진 형태 -> 무조건 2차함수.  
따라서 2차함수임이 확실하면, MLP가 삼각함수의 연산을 할 수 있다는 강력한 증거가 됨.  
<img width="638" height="400" alt="Image" src="https://github.com/user-attachments/assets/bba6dead-4b36-4f85-b2e0-1d5cea8f5a51" />

- 대부분의 뉴런이 2차함수 기능을 수행하고.  
- 주기 k=14 기준 $W_L$의 44 뉴런들을 뽑아서 봤을 때 sin, cos 의 k=14값만 활성 > 자기 역향을 충실하게 수행하고 있다..  
- 핵심 주기를 뺐을 때 오류가 급증한다.  

### 4. 억제를 통해 이게 진짜 핵심기능인지 파악하기 
<img width="666" height="364" alt="Image" src="https://github.com/user-attachments/assets/09e0cb59-00d0-4a9e-9d98-9d78830babb0" />  

핵심요소를 뺐을 때는 오류가 급증, 핵심요소를 제외한 나머지를 뺐을 때는 오히려 성능 좋아짐.

**Comment by Tonnonssi** [comment]
## 그로킹 이해하기  
두 오류 함수와 같이 $W_E, W_L$의 푸리에 구성요소의 norm을 갖는 the Gini coefficient도 정의. 얘는 각 가중치 벡터의 희소성을 푸리에 basis 안에서 특정한다.  

1. 암기 : 학습_두 오차함수 모두 감소하는 경향 테스트_the Gini coefficient는 평평하고, 두 오차함수는 높음  >> 걍 외운거고, 핵심 $w_k$는 아직 학습하지 못했다. 
2. 회로 형성 : e.l는 증가, 가중치의 ss는 감소하기 시작, test,train loss flat. > 암기한 해법이 푸리에 알고리즘화가 되고 있다. 가중치의 ss의 감소는 회로 구성이 가중치 감소를 통해 되고 있음을 시사. 그로킹이 시작되기 전에 일반화가 진행됨을 기억하자   
3. 클린업 : e.l 최고점, r.l 감소하기 시작, test loss 감소, weight의 ss도 크게 감소 >> 제대로 학습하면서 spares하게, 푸리에를 근사가 정교해짐
